### PR TITLE
Add Lion optimizer and a "skip step" version of it

### PR DESCRIPTION
**Pull request by epwalsh**
I don't really intend to use the Lion optimizer, this is more meant as a proof of concept for two things:
- an efficient (no host-device sync) implementation of the "skip step" optimizer idea that @2015aroras is investigating, and
- compiling the optimizer step

### CODE REVIEW

File: src/olmo_core/optim/skip_step_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        """
+        if len(self._losses) < max(20, self.rolling_interval_length // 2):
+            return torch.tensor(1.0).to(device=self.latest_loss.device, non_blocking=True)
+
+        loss_std = torch.std(torch.stack(self._losses[:-1]))
+
+        if self._grad_norms:
+            grad_norm_std = torch.std(torch.stack(self._grad_norms[:-1]))
+            return (
+                self.latest_loss <= self.sigma_factor * loss_std
```
Comment (2015aroras) [inline src/olmo_core/optim/skip_step_optimizer.py]: The Karpathy thing actually filters out when loss/grad norm is `sigma_factor` standard deviations above the mean (this does above 0). A number that is `s` standard deviations above the mean is said to have a "z-score" of `s`.

**Review by 2015aroras — COMMENTED** [review]
If I want to log the step_factor, do I lose all the benefits of avoiding host-device sync?

File: src/olmo_core/optim/skip_step_optimizer.py (hunk 1 above)
Comment (epwalsh) [reply src/olmo_core/optim/skip_step_optimizer.py]: Oh, duh, that's embarrassing after having spent 4 years in a stats PhD program

Comment (epwalsh) [reply src/olmo_core/optim/skip_step_optimizer.py]: 82d2c52

**Review by 2015aroras — APPROVED** [review]

File: src/olmo_core/optim/skip_step_optimizer.py (hunk 1 above)
Comment (epwalsh) [reply src/olmo_core/optim/skip_step_optimizer.py]: Forgot the `.abs()` but just pushed a fix for that.
